### PR TITLE
[#157431497] First draft o ADR-6: Backend is deployed on more than one instance

### PR DESCRIPTION
**Pull request by lussoluca**
First draft

### CODE REVIEW

File: doc/architecture/decisions/0006-backend-is-deployed-on-more-than-one-instance.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+We decided to use the Premium version that provides us the following features:
+
+* Redis data persistence
+* Redis cluster
+* Scale out to multiple cache units
+* Azure Virtual Network
+
+## Consequences
+
+Connections to the Redis cluster will be managed by the `redis-clustr`
```
Comment (cloudify) [inline doc/architecture/decisions/0006-backend-is-deployed-on-more-than-one-instance.md]: do keys get sharded across the instances? what happens to sharding when the cluster gets rescaled?

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Remember to fix the PR title by removing WIP wording when ready
  </td>
    </tr>
  </tbody>
</table>



## Affected stories

  * ⚙️ [#157431497](https://www.pivotaltracker.com/story/show/157431497): Scrivere un ADR per rendere il backend dell'app scalabile orizzontalmente

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: .env.example [hunk 2]
```diff
@@ -3,7 +3,9 @@ API_URL=https://agid-apim-prod.azure-api.net/api/v1
 CLIENT_REDIRECTION_URL=/profile.html?token={token}
 CLIENT_ERROR_REDIRECTION_URL=/error.html
 PORT=80
-REDIS_URL="redis://redis"
+REDIS_URL=put_the_azure_redis_url_here
+REDIS_PORT=put_the_azure_redis_port_here
+REDIS_PASSWORD=put_the_azure_redis_password_here
```
Comment (cloudify) [inline .env.example]: this should probably be passed by env var

File: src/container.ts [hunk 3]
```diff
@@ -183,7 +183,28 @@ container.register({
...
+    const redisUrl = process.env.REDIS_URL;
+    const redisPassword = process.env.REDIS_PASSWORD;
+    const redisPort: number = parseInt(
+      process.env.REDIS_PORT || DEFAULT_REDIS_PORT,
+      10
+    );
+
+    if (!redisUrl || !redisPassword || !redisPort) {
+      winston.error(
+        "Missing require environment variables needed to connect to Redis host (REDIS_URL, REDIS_PASSWORD, REDIS_PORT)"
```
Comment (cloudify) [inline src/container.ts]: required

File: .env.example (hunk 2 above)
Comment (lussoluca) [reply .env.example]: yes, the `dotenv` package reads from this file or from the environment variables (on k8s we set those values as environment variables).
In this way a user can run the proxy locally both on the host or in a Docker container

File: doc/architecture/decisions/0006-backend-is-deployed-on-more-than-one-instance.md [hunk 4]
```diff
@@ -21,7 +21,11 @@ We decided to use the Premium version that provides us the following features:
 * Scale out to multiple cache units
 * Azure Virtual Network
 
+From the external the cluster appears as a single connection so there is no need to use a Redis client that supports
```
Comment (gunzip) [inline doc/architecture/decisions/0006-backend-is-deployed-on-more-than-one-instance.md]: I've read that it needs a client library that supports clustering, where did you get this info ?

Comment (lussoluca) [reply doc/architecture/decisions/0006-backend-is-deployed-on-more-than-one-instance.md]: you are right, documentation is confusing, [here](https://docs.microsoft.com/en-us/azure/redis-cache/cache-how-to-premium-clustering) it says that "At the present time not all clients support Redis clustering", but after that it says "You can connect to your cache using the same endpoints, ports, and keys that you use when connecting to a cache that does not have clustering enabled. Redis manages the clustering on the backend so you don't have to manage it from your client.". Anyway the nodejs example [here](https://docs.microsoft.com/en-us/azure/redis-cache/cache-nodejs-get-started) uses the same node_redis package as we do.

File: doc/architecture/decisions/0006-backend-is-deployed-on-more-than-one-instance.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+We decided to use the Redis service provided by [Azure](https://azure.microsoft.com/en-us/services/cache).
+We decided to use the Premium version that provides us the following features:
+
+* Redis data persistence
+* Redis cluster
+* Scale out to multiple cache units
+* Azure Virtual Network
+
+From the external the cluster appears as a single connection so there is no need to use a Redis client that supports
```
Comment (gunzip) [inline doc/architecture/decisions/0006-backend-is-deployed-on-more-than-one-instance.md]: is this still valid ?

Comment (lussoluca) [reply doc/architecture/decisions/0006-backend-is-deployed-on-more-than-one-instance.md]: No, I'll fix it

**Pull request merged**